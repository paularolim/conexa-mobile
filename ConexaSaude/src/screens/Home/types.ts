import { RouteProp } from '@react-navigation/native';
import { NativeStackNavigationProp, NativeStackScreenProps } from '@react-navigation/native-stack';

import { Appointment } from '@hooks/useListAppointments';
import { TabsRoutesParamList } from '@routes/types';

export interface HomeViewProps {
  appointments: Appointment[] | null;
  name: string;
  refreshing: boolean;
  loading: boolean;
  error: boolean;
  handlePressAppointment: (id: number) => void;
  handleCreateAppointment: () => void;
  onRefresh: () => Promise<void>;
}

export type ScreenProps = NativeStackScreenProps<TabsRoutesParamList, 'Home'>;
export type NavigationType = NativeStackNavigationProp<TabsRoutesParamList, 'Home', undefined>;
export type RouteType = RouteProp<TabsRoutesParamList, 'Home'>;
