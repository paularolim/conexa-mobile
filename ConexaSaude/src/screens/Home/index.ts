export * from './Home';
export * from './HomeView';
export * from './types';
export * from './useHomeViewModel';
