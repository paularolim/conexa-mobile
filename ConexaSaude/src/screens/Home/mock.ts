export const appointments = {
  data: [
    {
      id: 16,
      medico: {
        id: 1,
        nome: 'Dr. Daniel Vieira',
        email: null,
      },
      paciente: 'Diego Senna',
      dataConsulta: '2020-04-23 08:30',
      observacao: 'Exemplo de consulta',
    },
    {
      id: 11,
      medico: {
        id: 1,
        nome: 'Dr. Daniel Vieira',
        email: null,
      },
      paciente: 'Rafael Braga',
      dataConsulta: '2020-04-23 08:30',
      observacao: 'Exemplo de consulta',
    },
    {
      id: 12,
      medico: {
        id: 1,
        nome: 'Dr. Daniel Vieira',
        email: null,
      },
      paciente: 'Igor Silva',
      dataConsulta: '2020-04-23 08:30',
      observacao: 'Exemplo de consulta',
    },
    {
      id: 13,
      medico: {
        id: 1,
        nome: 'Dr. Daniel Vieira',
        email: null,
      },
      paciente: 'Igor Costa',
      dataConsulta: '2020-04-23 08:30',
      observacao: 'Exemplo de consulta',
    },
    {
      id: 14,
      medico: {
        id: 1,
        nome: 'Dr. Daniel Vieira',
        email: null,
      },
      paciente: 'Deiveson Barreto',
      dataConsulta: '2020-04-23 08:30',
      observacao: 'Exemplo de consulta',
    },
    {
      id: 15,
      medico: {
        id: 1,
        nome: 'Dr. Daniel Vieira',
        email: null,
      },
      paciente: 'Adriano Sena',
      dataConsulta: '2020-04-23 08:30',
      observacao: 'Exemplo de consulta',
    },
  ],
};
