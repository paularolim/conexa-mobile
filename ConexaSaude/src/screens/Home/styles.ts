import { View } from 'react-native';
import styled from 'styled-components/native';

import { Text } from '@components/Text';

export const Container = styled(View)`
  flex: 1;
  padding: 24px 24px 0px 24px;
  gap: 20px;
  background-color: ${({ theme }) => theme.colors.white};
`;

export const Header = styled(View)`
  padding-bottom: 16px;
  gap: 8px;
  background-color: ${({ theme }) => theme.colors.white};
`;

export const Title = styled(Text)`
  font-size: 20px;
  font-weight: bold;
  color: ${({ theme }) => theme.colors.gray};
`;

export const Separator = styled(View)`
  height: 10px;
`;

export const Footer = styled(View)`
  height: 30px;
`;
