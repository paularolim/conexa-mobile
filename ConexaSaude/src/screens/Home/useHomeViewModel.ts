import { useCallback, useEffect, useState } from 'react';

import { useAuth } from '@hooks/useAuth';
import { useListAppointments } from '@hooks/useListAppointments';

import { ScreenProps } from './types';

export function useHomeViewModel({ navigation }: ScreenProps) {
  const [refreshing, setRefreshing] = useState(false);

  const { getAppointments, appointments, loading, error } = useListAppointments();
  const { name } = useAuth();

  const requestAppointments = useCallback(() => {
    getAppointments();
  }, [getAppointments]);

  const onRefresh = useCallback(async () => {
    setRefreshing(true);
    await getAppointments();
    setRefreshing(false);
  }, [getAppointments]);

  const handlePressAppointment = useCallback(
    (id: number) => {
      navigation.navigate('Appointment', { id });
    },
    [navigation],
  );

  const handleCreateAppointment = () => {
    navigation.navigate('CreateAppointment');
  };

  useEffect(() => requestAppointments(), [requestAppointments]);

  return {
    handleCreateAppointment,
    handlePressAppointment,
    refreshing,
    onRefresh,
    name: name || '',
    appointments,
    loading,
    error,
  };
}
