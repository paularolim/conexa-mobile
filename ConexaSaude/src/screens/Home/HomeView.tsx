import { useCallback } from 'react';
import { FlatList, ListRenderItemInfo, RefreshControl } from 'react-native';

import { AppointmentCard } from '@components/AppointmentCard';
import { Appointment } from '@hooks/useListAppointments/types';

import { Container, Footer, Header, Separator, Title } from './styles';
import { HomeViewProps } from './types';

export function HomeView({
  name,
  refreshing,
  onRefresh,
  appointments,
  handlePressAppointment,
}: HomeViewProps) {
  const headerItem = useCallback(
    () => (
      <Header>
        <Title>Ol√° {name}!</Title>
        <Title>Agenda</Title>
      </Header>
    ),
    [name],
  );

  const footerItem = useCallback(() => <Footer />, []);

  const separatorItem = useCallback(() => <Separator />, []);

  const renderItem = useCallback(
    ({ item }: ListRenderItemInfo<Appointment>) => (
      <AppointmentCard date={item.date} patient={item.patient} onPress={handlePressAppointment} />
    ),
    [handlePressAppointment],
  );

  return (
    <Container>
      <FlatList
        data={appointments || []}
        keyExtractor={(item) => `${item.id}`}
        renderItem={renderItem}
        ItemSeparatorComponent={separatorItem}
        showsVerticalScrollIndicator={false}
        stickyHeaderIndices={[0]}
        ListHeaderComponent={headerItem}
        ListFooterComponent={footerItem}
        refreshControl={<RefreshControl refreshing={refreshing} onRefresh={onRefresh} />}
      />
    </Container>
  );
}
