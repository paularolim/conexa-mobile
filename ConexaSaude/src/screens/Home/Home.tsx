import { HomeView } from './HomeView';
import { ScreenProps } from './types';
import { useHomeViewModel } from './useHomeViewModel';

export function Home(screenProps: ScreenProps) {
  const props = useHomeViewModel(screenProps);
  return <HomeView {...props} />;
}
