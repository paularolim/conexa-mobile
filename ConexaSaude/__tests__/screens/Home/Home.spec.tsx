import renderer from 'react-test-renderer';

import { Home, NavigationType, RouteType, ScreenProps } from '@screens/Home';

import { Wrapper } from '../../Wrapper';

jest.mock('@screens/Home/useHomeViewModel', () => ({ useHomeViewModel: jest.fn() }));

const createTestProps = (): unknown & ScreenProps => ({
  navigation: { navigate: jest.fn() } as unknown as NavigationType,
  route: { params: { id: '1' } } as unknown as RouteType,
});
const props = createTestProps();

describe('Home', () => {
  it('should render screen correctly', () => {
    const tree = renderer
      .create(
        <Wrapper>
          <Home navigation={props.navigation} route={props.route} />
        </Wrapper>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
