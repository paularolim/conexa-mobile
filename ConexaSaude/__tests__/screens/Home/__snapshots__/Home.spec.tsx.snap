// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render screen correctly 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "#FFFFFF",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "gap": 20,
        "paddingBottom": 0,
        "paddingLeft": 24,
        "paddingRight": 24,
        "paddingTop": 24,
      },
    ]
  }
>
  <RCTScrollView
    ItemSeparatorComponent={[Function]}
    ListEmptyComponent={[Function]}
    ListFooterComponent={[Function]}
    ListHeaderComponent={[Function]}
    data={[]}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    refreshControl={<RefreshControlMock />}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={
      [
        0,
      ]
    }
    viewabilityConfigCallbackPairs={[]}
  >
    <RCTRefreshControl />
    <View>
      <View
        collapsable={false}
        onLayout={[Function]}
      >
        <View
          style={
            [
              {
                "backgroundColor": "#FFFFFF",
                "gap": 8,
                "paddingBottom": 16,
              },
            ]
          }
        >
          <Text
            style={
              [
                {
                  "color": "#000000",
                },
                {
                  "color": "#707683",
                  "fontSize": 20,
                  "fontWeight": "bold",
                },
              ]
            }
          >
            Ol√° 
            !
          </Text>
          <Text
            style={
              [
                {
                  "color": "#000000",
                },
                {
                  "color": "#707683",
                  "fontSize": 20,
                  "fontWeight": "bold",
                },
              ]
            }
          >
            Agenda
          </Text>
        </View>
      </View>
      <View
        style={
          [
            {
              "gap": 8,
              "paddingBottom": 24,
              "paddingLeft": 0,
              "paddingRight": 0,
              "paddingTop": 24,
            },
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#000000",
              },
              {
                "color": "#000000",
                "fontSize": 18,
              },
            ]
          }
          testID="empty-message"
        >
          Sem consultas para exibir. Que tal criar uma?
        </Text>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={false}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#2ED47A",
              "borderRadius": 4,
              "justifyContent": "center",
              "maxWidth": "100%",
              "minHeight": 48,
              "opacity": 1,
              "paddingBottom": 8,
              "paddingLeft": 16,
              "paddingRight": 16,
              "paddingTop": 8,
            }
          }
        >
          <Text
            style={
              [
                {
                  "color": "#000000",
                },
                {
                  "color": "#FFFFFF",
                  "fontSize": 18,
                },
              ]
            }
            type="filled"
          >
            Criar consulta
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            [
              {
                "height": 30,
              },
            ]
          }
        />
      </View>
    </View>
  </RCTScrollView>
</View>
`;
